text/0000-explode.md

```diff
@@ -0,0 +1,236 @@
+
+# How We Teach This
+
+The most important message we need to teach app developers is: use `ember-cli-update` whenever you're changing your Ember version. As long as we can spread that message, we can provide direct guidance the rest of the way. For example, when we reach a sufficient level of confidence in the `@ember/core` packaging, we can begin making `ember-cli-update` offer to automatically explode `ember-source` into `@ember/core` *et al*. This would be a simple code mod that should not alter any app semantics.
+
+As people begin to experiment with removing Ember packages, our best teaching opportunities are very clear and helpful feedback from `ember-cli` whenever they have a dependency issue, as illustrated in some of the example messages in this document.
+
+The mental model for "which Ember packages am I using?" should ultimately reduce to simple rules:
+
+ - in Javascript, you can see directly from your ES imports which packages you are depending on.
+ - in templates you will often be able to see what package a component or helper comes from thanks to [RFC 143](https://github.com/emberjs/rfcs/blob/master/text/0143-module-unification.md), which adds namespacing.
+ - Ember-provided components and helpers don't necessarily require namespaces, but we can offer precise guidance if you use one that isn't installed, like "No {{link-to}} component found. You may need to install @ember/routing."
+
+We don't need to make any immediate changes to the guides, because the default behavior of Ember isn't changing. And even if people copy code examples out of more "barebones" Ember apps, all those examples will work in a default (complete, `ember-source`-based) app.
+
+As the barebones app experiences (described above under New App Blueprint) matures, we should develop a separate guide designed around how to build apps with that starting point. 
+
+# Drawbacks
+
+The biggest drawback of Ember-as-many-packages is that NPM's peerDependency support is relatively weak. Simply saying `npm install @ember/foo` won't automatically take peerDependencies into account to choose the precise version of `@ember/foo` that matches your other packages. We can easily tell you when your peerDeps are wrong, but we can't always easily fix them for you automatically, unless we start laying specialized tools on top of NPM.
+
+In the case of code we want to remove, splitting it into separate packages doesn't eliminate our need to support it (at least until the next major release lets it be truly removed). So while this proposal lets us "go faster" in some senses, it doesn't let us reduce support any faster.
+
+# Alternatives
+
+We could continue to ship Ember as a single package, and do our own Ember-specific built-time shenanigans to decide how much of it to include in the app's build. This greatly simplifies the management of user's package.json files, but it is increasingly out-of-step with wider NPM-based Javascript toolchains.
```

Comment:
Abandoning the "npm-SemVer" model won't help to keep in step with wider NPM-based Javascript toolchains.

Reply:
I don't think it's correct to say that this proposal "abandons the npm semver model".

In general, this proposal uses `peerDependencies` in the usual way. It also creates a new API (like any other API), which allows addons to ask to opt-out of features they don't want to use.

Declaring a baseline version is basically just saying "I want all of the features that were on by default in 3.5", and then the package opts out of additional features known at the time. That means that those same features will still be active in 3.6 without any changes.

In a future version, if an addon wants to drop additional features "by default", it can increment its baseline. This is not in **conflict** with npm's story, and it doesn't change anything about semantic versioning. It's just expressing something a little more subtle, but if people ignore the subtlety, the only consequence is that they pull in more code (which is the status quo).

Reply:
In `npm`'s story I could just increase the minor version of `@ember/core`, `npm install`, and everything should work. This proposal abandons this model with the need to use special tooling outside the scope of regular `npm`-ed packages. This is what I was referring to.

Reply:
That isn't exactly correct, because if you have multiple packages that depend on each other, you'd also have to bump the other packages manually. This tooling basically helps with that manual bit.

Reply:
@knownasilya Could you give an example where - given packages with regurlar npm/SemVer semantics - I cannot bump just one minor version number in my `package.json` without bumping others? My understanding is, that the whole point of SemVer is that I'm allowed to do that.

Reply:
@jnfingerle as written, I've proposed making the separate packages have exact peerDependencies on each other. So if `@ember/routing` needs `@ember/core`, it would express that via an exact version peerDependency.

Over time we're free to relax those constraints to use `^` or `~` dependencies instead, which would enable increased mixing and matching. But the simplest thing for both consumers and contributors is to start with the most exact constraint first.

This is all just standard npm stuff now (the latest version of the RFC removed the special upgrade requirement).

Reply:
@jnfingerle also I wrote the above comment before I saw that you've basically already said the same thing at the bottom of the thread. Sounds like we are approaching consensus.